2021-year-report.org

```diff
@@ -0,0 +1,438 @@
+
+See: [[https://doc.rust-lang.org/reference/paths.html#qualified-paths][QualifiedPathInType]]
+
+**** Never Type
+
+Rust has a never type which is pretty interesting. In c/c++, you can use the attribute no_return to signify that this function is never going to return. In Rust, they take this concept to its full with the never type, so for example, below: 
+
+#+BEGIN_SRC rust
+fn test() {
+   let a = return;
+   let b = a + 123;
+   a = 456;
+}
+#+END_SRC
+
+You can see 'a' is equal to the return expression, which will, in this context, never return. So even though everything after this, the first assignment will never get executed because of the return expression, Rust can fully type this function.  The functionality here is down to a key piece of the type system, which is more evident in closures than here. Still, in the above 'b = a + 123', the types here never type plus an integer inference variable, as there is no implementation of never type plus integer, here Rust actually returns another inference variable with a specified type-bound that requires the ability to add numbers together. This means the final assignment allows the type system to know that 'a' is an integer, albeit an uninitialized one, but this code is still safe because of the return expression.
+
+**** Method resolution 
+
+Method resolution is a multi-variant task; it involves the autoderef mechanism, but more importantly, there is an implicit ordering of the candidates that is key as plain old Impl block methods are always prefered over trait's. 
+
+#+BEGIN_SRC rust
+struct Foo {}
+
+trait Bar where Self:Sized {
+  fn bar(self) {
+      println!("In trait def!")
+  }
+}
+
+impl Foo {
+  fn bar(self) {
+    println!("In struct impl!")
+  }
+}
+
+impl Bar for Foo {
+  fn bar(self) {
+    println!("In trait impl!")
+  }
+}
+
+fn main() {
+  let mut f = Foo{};
+  f.bar();
+}
+#+END_SRC
+
+This test case is all about showing even though there are two duplicate methods here, we enforce that the impl block method is prefered first.
+
+see:
+https://rustc-dev-guide.rust-lang.org/method-lookup.html
+https://doc.rust-lang.org/nightly/nomicon/dot-operator.html
+
+**** Add check for duplicate overlapping impl-items
+
+Rust allows multiple impl blocks to give a generic data type specialization. But suppose the programmer adds a generic impl for a duplicate method. In that case, it will become impossible to distinguish the method 'bar' in each of these specialized impl blocks for method resolution. Since your receiver could be Foo<_ (inference variable)> which could resolve to Foo<isize> or Foo<char>. see: rustc --explain E0592
```

Comment:
I don't really get what the point of this example is. That there's no specialization in Rust (yet)?

Reply:
Its hard to explain but adding the generic impl block is an error that you can catch at type-checking not at the name-resolution error even though its a duplicate bla found.